### PR TITLE
AddBlock refactor

### PR DESCRIPTION
**Pull request by dzucconi**
Closes https://github.com/aredotna/ervell/issues/612

add-block

add-block-list

### CODE REVIEW

**Comment by dzucconi** [comment]
For some reason gifs wash out greys. But we'll peep it on staging.

File: apps/channel/client/block_skeleton_view.coffee [hunk 1]
```diff
@@ -1,179 +0,0 @@
...
-    _.each @pages, (page)->
-      return false unless page.loaded
-
-    return true
-
-  cancelRequest: (id)->
-    @pages[id]?.xhr?.abort?()
-    @pages[id]?.loaded = false
-    @pages[id]?.loading = false
-    @unsetPlaceholdersLoading(id)
```
Comment (dzucconi) [inline apps/channel/client/block_skeleton_view.coffee]: This didn't appear to be used anywhere.

File: apps/channel/client/index.coffee [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    blocks: blocks
+    blockCollectionView: view
+    resultsCollection: resultsCollection
+
+  initChannelPath channel
+
+  if current_user.canAddToChannel(channel) and not (FOLLOWERS? or FOLLOWING?)
+    addBlockView = AddBlockView $('.js-add-block'), blocks
+
+    resultsCollection?.on 'reset', ->
```
Comment (dzucconi) [inline apps/channel/client/index.coffee]: I don't entirely grok why this is needed but it seems to be.

File: apps/channel/client/block_skeleton_view.coffee (hunk 1 above)
Comment (broskoski) [inline apps/channel/client/block_skeleton_view.coffee]: Yes, you're right

**Comment by dzucconi** [comment]
I forgot to delete the old new block code. For some good git line diff drama.

File: apps/channel/client/index.coffee (hunk 2 above)
Comment (broskoski) [inline apps/channel/client/index.coffee]: When a user types in the filter and then erases, the resultsCollection needs to get cleared and the addBlock button needed to get re-initialized.

File: apps/statuses/index.coffee [hunk 3]
```diff
@@ -0,0 +1,18 @@
+app = module.exports = require('express')()
+
+statuses = [
```
Comment (broskoski) [inline apps/statuses/index.coffee]: What's this?

File: components/add_block/client/view.coffee [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      button: @$('.js-button')
+
+    @checkInput()
+
+  render: ->
+    @$el.html template()
+
+    @postRender()
+
+    this
```
Comment (broskoski) [inline components/add_block/client/view.coffee]: Much nicer 👌

File: apps/statuses/index.coffee (hunk 3 above)
Comment (dzucconi) [inline apps/statuses/index.coffee]: Just what I was developing against. So I can change the url to always fail or succeed and make sure states work as they should.

File: apps/channel/client/index.coffee [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  channelView = new ChannelView
+    el: $('.js-channel')
+    channel: channel
+    blocks: blocks
+    blockCollectionView: view
+    resultsCollection: resultsCollection
+
+  initChannelPath channel
+
+  if current_user.canAddToChannel(channel) and not (FOLLOWERS? or FOLLOWING?)
```
Comment (broskoski) [inline apps/channel/client/index.coffee]: Ok, this took a second to figure out, but there's going to be an error with this setup. 

The reason the new block view is getting initialized inside the actual `ChannelView` up there is this line:
```coffeescript
mediator.on 'collaborators:fetched', @checkUserAbilities, @
```
Channels where the user is a collaborator do not come down with every member of the channel, it has to be fetched after the fact.

Comment (broskoski) [inline apps/channel/client/index.coffee]: To clarify, this addBlock view will not show up now on channels where a user is a collaborator (and not the owner).

Comment (dzucconi) [inline apps/channel/client/index.coffee]: This is code that was copied from the existing implementation. But I'll read it now

File: apps/channel/client/index.coffee [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  initChannelPath channel
+
+  # Add Block initialization:
+  if current_user.canAddToChannel(channel) and not (FOLLOWERS? or FOLLOWING?)
+    integration = addBlockChannelServerRenderedIntegration
+      $el: $('.js-add-block')
+      collection: blocks
+
+    integration.init()
+    resultsCollection?.on 'reset', integration.reset
```
Comment (dzucconi) [inline apps/channel/client/index.coffee]: Default integration, now wrapped up into it's own thing.

File: apps/channel/client/index.coffee [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  mediator.once 'collaborators:fetched', (collaborators) ->
+    return if $('.js-add-block').length # Already rendered
+    return unless mediator.shared.current_user.id in collaborators.pluck 'id'
+
+    integration = addBlockChannelClientRenderedIntegration
+      $el: $('.js-block-collection')
+      collection: blocks
+
+    integration.init()
+    resultsCollection?.on 'reset', integration.reset
```
Comment (dzucconi) [inline apps/channel/client/index.coffee]: Client-side integration, similarly

**Comment by dzucconi** [comment]
All should be well

**Comment by broskoski** [comment]
👍

**Pull request merged**